src/lib/reasoners/ac.ml

```diff
@@ -163,26 +163,61 @@ module Make (X : Sig.X) = struct
   let fold_flatten sy f =
     List.fold_left (fun z (rt,n) -> flatten sy ((f rt),n) z) []
 
-  let abstract2 sy t r acc =
+  let is_other_ac_symbol sy r =
     match X.ac_extract r with
-    | Some ac when Sy.equal sy ac.h -> r, acc
-    | None -> r, acc
-    | Some _ -> match Expr.term_view t with
-      | { Expr.f = Sy.Name { hs; kind = Sy.Ac; _ }; xs; ty; _ } ->
+    | Some ac -> not (Sy.equal sy ac.h)
+    | None -> false
+
+  (* This implements a variant of the term abstraction process described in
+     section 6 of the AC(X) paper [1].
+
+     The abstraction process given in the paper requires to abstract all AC
+     symbols appearing in a non-AC context, but the implementation does not
+     know about the context at the time the abstraction is performed (note that
+     rules Abstract1 and Abstract2 are concerned with *equations* while the
+     abstraction process here occurs at the time of building semantic values,
+     which is earlier). Further, the implementation seems to implicitly rely on
+     term ordering (older terms are ordered before newer terms, so in
+     particular subterms are always smaller than terms that contains them) to
+     cheaply prevent loops rather than introducing all the abstracted variables
+     that the theoretical presentation in the paper would require.
+
+     So the implementation below of the Abstract2 rules deviates from the
+     presentation in the paper to accomodate those differences, globally,
+     between the implementation and theoretical description of AC(X).
+
+     More precisely, `abstract2` will abstract terms that *contain* AC leaves
+     when they appear as argument of an AC symbol. This ensures that AC terms
+     satisfy the T_AC definition from page 22 of the paper, although
+     correctness of the corresponding abstraction process has not been proven.
+     See also https://github.com/OCamlPro/alt-ergo/issues/989
+
+     [1]: https://arxiv.org/pdf/1207.3262.pdf *)
+  let abstract2 sy t r acc =
+    if List.exists (is_other_ac_symbol sy) (X.leaves r) then
```

Comment:
I  guess this is the only important line in this PR :)

Let's `u` and `v` two distinct AC symbols (for sake of simplicity, say that `u` and `v` are binary integer operators).
Consider the expressions `A := u(v(x, y), z)` and `B := u(v(x, y) + t, z)`. 

In the previous implementation, we abstracted the subterm `v(x, y)` in the former case but we don't in the latter one. 

In the new implementation, we'll abstract both of them because `v(x, y)` is now a leaf of an argument of `u` and `u` is distinct of `v`. 

If we replace the operator `+` by a non-ac uninterpreted symbol, we don't abstract the subterm `v(x, y)`, right? For instance we don't abstract `v(x, y)` in the term `u(h(v(x, y), t), z)` where `h` is a non-ac uninterpreted operator.

I guess the notation Σ_∅ in 𝒯_∅ in the paper denotes the interpreted symbols? 

We don't need to abstract uninterpreted symbols here because, by definition they are already abstracted, so there is no relation between `h(x, y)` and `h(y, x)` that can induce an infinity path in the reduction algorithm?